File: tps/entrada e saida/tp03/TP03Q07 - Fila Circular com Alocação Flexível em C/jogador.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>
#include <math.h>

#define TAM_MAX_LINHA 250
#define MAX_ENTRADAS 250
#define arquivoCSV "/tmp/players.csv"

typedef struct {
  int id;
  int peso;
  int altura;
  char nome[70];
  char universidade[70];
  int anoNascimento;
  char cidadeNascimento[70];
  char estadoNascimento[70];
} Jogador;


void setJogador(char* strAttr[], Jogador* J) {
  J->id = atoi(strAttr[0]);
  strcpy(J->nome, strAttr[1]);
  J->altura = atoi(strAttr[2]);
  J->peso = atoi(strAttr[3]);
  strcpy(J->universidade, strAttr[4]);
  J->anoNascimento = atoi(strAttr[5]);
  strcpy(J->cidadeNascimento, strAttr[6]);
  strcpy(J->estadoNascimento, strAttr[7]);
}


void imprimir(Jogador *J) {
  char* vazio = "nao informado";
  printf("## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
    //J->id,
    J->nome[0]=='\0' ? vazio : J->nome,
    J->altura,
    J->peso,
    J->anoNascimento,
    J->universidade[0]=='\0' ? vazio : J->universidade,
    J->cidadeNascimento[0]=='\0' ? vazio : J->cidadeNascimento,
    J->estadoNascimento[0]=='\0' ? vazio : J->estadoNascimento
  );
}


Jogador clone(Jogador *jogador) {
  Jogador retorno;
  retorno.id = jogador->id;
  strcpy(retorno.nome, jogador->nome);
  retorno.altura = jogador->altura;
  retorno.peso = jogador->peso;
  retorno.anoNascimento = jogador->anoNascimento;
  strcpy(retorno.universidade, jogador->universidade);
  strcpy(retorno.cidadeNascimento, jogador->cidadeNascimento);
  strcpy(retorno.estadoNascimento, jogador->estadoNascimento);
  return retorno;
}


// recebe string de uma lihna do csv e coloca os atributos no array
void parseAtributos(char* input, char* output[]) {
  int i = 0;
  strtok(input, "\n"); // retira quebra de linha no final da string
  
  output[i++] = input;

  // loop vai repartindo linha nas ',' e atribuindo a output[]
  while (i < 8 && input && (input = strchr(input, ','))) {
    *input = 0;           
    output[i++] = ++input;
  }
}


// le a lihna do csv correspondente ao id procurado
void lerLinha(FILE* csv, char linha[], int id) {
  rewind(csv);
  for(int n = 0; n <= id + 1; n++) {
    fgets(linha, TAM_MAX_LINHA, csv);
  }
}


// atribui cada atributo do array de strings no objeto Jogador
Jogador ler(FILE *csv, int id) {
  Jogador J;
  char linha[TAM_MAX_LINHA]; // linha a ser lida do csv 
  char* strAttr[8]; // array com os atributos

  lerLinha(csv, linha, id); // le linha do arquivo correspondente ao id
  
  parseAtributos(linha, strAttr); // transforma a string em array de atributos
  
  setJogador(strAttr, &J); // cria o registro jogador a partir do array de atributos
  
  return J; 
}


// faz leitura dos valores a partir da entrada padrao e 
// retorna numero de entrada efetuadas
int leituraDeEntrada(char entrada[][TAM_MAX_LINHA]) {
  int numEntrada = 0; // contador de entradas
  do {
    scanf("%[^\n]%*c", entrada[numEntrada]); // *c ignora ultimo caracter lido
  } while(strcmp(entrada[numEntrada++],"FIM")!=0);
  numEntrada--;  // desconsidera entrada com FIM
  return numEntrada;
}



// CELULA ============================================
// celula composta de objeto Jogador e ponteiro para o proximo Jogador
typedef struct Celula {
	Jogador elemento;        // Elemento inserido na celula.
	struct Celula* prox; // Aponta a celula prox.
} Celula;

Celula* novaCelula(Jogador* elemento) {
  Celula* nova = (Celula*) malloc(sizeof(Celula));
  nova->elemento = *elemento;
  nova->prox = NULL;
  return nova;
}

//FILA PROPRIAMENTE DITA =============================
Celula* primeiro;
Celula* ultimo;

// Cria uma fila sem elementos (somente no cabeca).
void start () {
  Jogador j;
  primeiro = novaCelula(&j);
  ultimo = primeiro;
}

// Insere elemento na fila (politica FIFO).
// @param x int Elemento a inserir.
void inserir(Jogador* x) {
  ultimo->prox = novaCelula(x);
  ultimo = ultimo->prox;
}

// Remove elemento da fila (politica FIFO).
// @return Elemento removido.
Jogador remover() {
  if (primeiro == ultimo) {
    errx(1, "Erro ao remover!");
  }
  Celula* tmp = primeiro;
  primeiro = primeiro->prox;
  Jogador resp = primeiro->elemento;
  tmp->prox = NULL;
  free(tmp);
  tmp = NULL;
  return resp;
}

// Mostra os elementos separados por espacos.
void mostrar() {
  Celula* i;
  int cont = 0;
  for (i = primeiro->prox; i != NULL; i = i->prox) {
    printf("[%d] ", cont);
    imprimir(&i->elemento);
    cont++;
  }
}

// Calcula e retorna o tamanho, em numero de elementos, da lista.
int tamanho() {
  Celula* i;
  int cont = 0;
  for (i = primeiro->prox; i != NULL; i = i->prox) {
    cont++;
  }
  return cont;
}

// Calcula media aredondada das alturas dos jogadores contidos na lista
int calculaMedia() {
  double media = 0;
  Celula* i;
  for (i = primeiro->prox; i != NULL; i = i->prox) {
    media += i->elemento.altura;
  }
  media /= tamanho();
  return round(media);
}



int main(int argc, char *argv[]) {
  FILE *csv;
  csv = fopen(arquivoCSV, "r");

  // variavel para armazenar entrada padrao com id's
  char entrada[MAX_ENTRADAS][TAM_MAX_LINHA]; 
  
  // leitura da entrada padrao, retorna quantidade de entradas inseridas
  int numEntrada = leituraDeEntrada(entrada); 
  
  start();
  
  // criar registros "jogador" a partir das entradas armazenadas
  for (int i = 0; i < numEntrada; i++) {
    Jogador j = ler(csv, atoi(entrada[i])); // atoi(entrada[i]) = id
    // inserir cada registro no fim da lista
    if(tamanho()>=5) {
      remover();
      inserir(&j); 
    } else {
      inserir(&j);
    }
    printf("%d\n", calculaMedia());
  }

  int numOperacoes, posicao, id; 
  scanf("%d",&numOperacoes);
  
  for(int i = 0; i<numOperacoes; i++) {
    Jogador jogador;
    char operacao[2]; scanf("%s", operacao);

      //operacao inserir
      if (strcmp(operacao,"I") == 0){ 
        scanf("%d", &id); 
        jogador = ler(csv, id);
        if(tamanho()>=5) { // checa tamanho da fila
          remover();
          inserir(&jogador); 
        } else {
          inserir(&jogador);
        }
        printf("%d\n", calculaMedia());
      }

      // operacao remover
      else if (strcmp(operacao,"R") == 0){
        printf("(R) %s\n", primeiro->prox->elemento.nome);
        remover(); 
      }
      else printf("operacao invalida\n");

  }
    
  // imprime lista de registros "jogador"
  mostrar();

  fclose(csv);
  return 0;
  
}

File: tps/entrada e saida/tp03/TP03Q06 - Pilha com Alocação Flexível em Java/Jogador.java
public class Jogador {

  static final String arquivoCSV = "/tmp/players.csv";
  static final int maxEntradas = 128;
  
  // celula composta de objeto Jogador e ponteiro para o proximo Jogador
  static class Celula {
    public Jogador elemento; // Elemento inserido na celula.
    public Celula prox; // Aponta a celula prox.
    public Celula() {
      this.elemento = null;
      this.prox = null;
    }
    public Celula(Jogador elemento) {
        this.elemento = elemento;
        this.prox = null;
    }
  }

  static class Lista {
    private Celula primeiro;
    private Celula ultimo;


    public Lista() {
      primeiro = new Celula();
      ultimo = primeiro;
    }


    public void inserirInicio(Jogador x) {
      Celula tmp = new Celula(x);
      tmp.prox = primeiro.prox;
      primeiro.prox = tmp;
      if (primeiro == ultimo) {           
        ultimo = tmp;
      }
        tmp = null;
    }


    public void inserirFim(Jogador x) {
      ultimo.prox = new Celula(x);
      ultimo = ultimo.prox;
    }


    public Jogador removerInicio() throws Exception {
      if (primeiro == ultimo) {
        throw new Exception("Erro ao remover (vazia)!");
      }

      Celula tmp = primeiro;
      primeiro = primeiro.prox;
      Jogador resp = primeiro.elemento;
      tmp.prox = null;
      tmp = null;

      System.out.println("(R) " + resp.getNome());
      return resp;
    }    


    public Jogador removerFim() throws Exception {
      if (primeiro == ultimo) {
        throw new Exception("Erro ao remover (vazia)!");
      } 

      // Caminhar ate a penultima celula:
      Celula i;
      for(i = primeiro; i.prox != ultimo; i = i.prox);

      Jogador resp = ultimo.elemento; 
      ultimo = i; 
      i = ultimo.prox = null;

      System.out.println("(R) " + resp.getNome());  
      return resp;
    }


    public void inserir(int pos, Jogador x) throws Exception {

        int tamanho = tamanho();

        if(pos < 0 || pos > tamanho){
        throw new Exception("Erro ao inserir posicao (" + pos + " / tamanho = " + tamanho + ") invalida!");
        } else if (pos == 0){
          inserirInicio(x);
        } else if (pos == tamanho){
          inserirFim(x);
        } else {
        // Caminhar ate a posicao anterior a insercao
          Celula i = primeiro;
          for(int j = 0; j < pos; j++, i = i.prox);
      
          Celula tmp = new Celula(x);
          tmp.prox = i.prox;
          i.prox = tmp;
          tmp = i = null;
        }
    }


    public Jogador remover(int pos) throws Exception {
      Jogador resp;
      int tamanho = tamanho();

      if (primeiro == ultimo){
        throw new Exception("Erro ao remover (vazia)!");

        } else if(pos < 0 || pos >= tamanho) {
          throw new Exception("Erro ao remover (posicao " + pos + " / " + tamanho + " invalida!");
        } else if (pos == 0){
          resp = removerInicio();
        } else if (pos == tamanho - 1){
          resp = removerFim();
        } else {
        // Caminhar ate a posicao anterior a insercao
          Celula i = primeiro;
          for(int j = 0; j < pos; j++, i = i.prox);
      
          Celula tmp = i.prox;
          resp = tmp.elemento;
          i.prox = tmp.prox;
          tmp.prox = null;
          i = tmp = null;
        }
      System.out.println("(R) " + resp.getNome());
      return resp;
    }    


    public void mostrar() {
      int j = 0;
      for (Celula i = primeiro.prox; i != null; i = i.prox) {
        System.out.print("[" + (j++) + "] ");
        i.elemento.imprimir();
      }
    }


    public boolean pesquisar(Jogador x) {
      boolean resp = false;
      for (Celula i = primeiro.prox; i != null; i = i.prox) {
          if(i.elemento.getNome().contains(x.getNome())){
              resp = true;
              i = ultimo;
          }
      }
      return resp;
    }


    public int tamanho() {
        int tamanho = 0; 
        for(Celula i = primeiro; i != ultimo; i = i.prox, tamanho++);
        return tamanho;
    }
  }


  int id;
  String nome;
  int altura;
  int peso;
  String universidade;
  int anoNascimento;
  String cidadeNascimento;
  String estadoNascimento;

  public Jogador() {
  }

  public Jogador(String linha) {
    String campos[] = linha.split(",");
    this.id = Integer.parseInt(campos[0]);
    this.nome = campos[1];
    this.altura = Integer.parseInt(campos[2]);
    this.peso = Integer.parseInt(campos[3]);
    this.universidade = (campos[4].isEmpty()) ? "nao informado" : campos[4];
    this.anoNascimento = Integer.parseInt(campos[5]);
    if (campos.length > 6) {
      this.cidadeNascimento = (campos[6].isEmpty()) ? "nao informado": campos[6];
      if (campos.length < 8) {
        this.estadoNascimento = "nao informado";
      } else {
        this.estadoNascimento = campos[7];
      }
    } else {
      this.cidadeNascimento = "nao informado";
      this.estadoNascimento = "nao informado";
    }
  }

  // id,Player,height,weight,collage,born,birth_city,birth_state
  public int getId() {
    return id;
  }

  public void setId(int id) {
    this.id = id;
  }

  public String getNome() {
    return nome;
  }

  public void setNome(String nome) {
    this.nome = nome;
  }

  public int getAltura() {
    return altura;
  }

  public void setAltura(int altura) {
    this.altura = altura;
  }

  public int getPeso() {
    return peso;
  }

  public void setPeso(int peso) {
    this.peso = peso;
  }

  public void setAnoNascimento(int anoNascimento){
    this.anoNascimento = anoNascimento;
  }

  public int getAnoNascimento(){
    return anoNascimento;
  }

  public String getUniversidade() {
    return universidade;
  }

  public void setUniversidade(String universidade) {
    this.universidade = universidade;
  }

  public String getCidadeNascimento() {
    return cidadeNascimento;
  }

  public void setCidadeNascimento(String cidadeNascimento) {
    this.cidadeNascimento = cidadeNascimento;
  }

  public String getEstadoNascimento() {
    return estadoNascimento;
  }

  public void setEstadoNascimetno(String estadoNascimento) {
    this.estadoNascimento = estadoNascimento;
  }

  public Jogador clone() {
    Jogador novo = new Jogador();
    novo.id = this.id;
    novo.nome = this.nome;
    novo.altura = this.altura;
    novo.peso = this.peso;
    novo.anoNascimento = this.anoNascimento;
    novo.universidade = this.universidade;
    novo.cidadeNascimento = this.cidadeNascimento;
    novo.estadoNascimento = this.estadoNascimento;
    return novo;
  }

  public void imprimir() {
    System.out.println("## " + nome + " ## " + altura + " ## " + peso + " ## " + anoNascimento + " ## "
        + universidade + " ## " + cidadeNascimento + " ## " + estadoNascimento + " ##");
  }

  public String toString() {
    return "[" + id + " ## " + nome + " ## " + altura + " ## " + peso + " ## " + anoNascimento + " ## "
      + universidade + " ## " + cidadeNascimento + " ## " + estadoNascimento + "]";
  }

  
  // retorna String com conteudo da linha correspondente ao 
  // id enviado como parametro
  public static String leLinhaCSV(int id) {
    Arq.openRead(arquivoCSV);
    String resp = "";
    // mover cursor do csv ate a linha do id
    for(int n = 0; n <= id + 1; n++) { 
      resp = Arq.readLine();  // le a linha com id inserido e armazena a string em "linha"
    }
    Arq.close();
    return resp;
  }


  // preenche vetor enviado como parametro com leitura de entrada padrao
  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(int entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readInt();
    } while (entrada[numEntrada++] != -1);
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }


  // preenche vetor enviado como parametro com leitura de entrada padrao
  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(String entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readLine();
    } while (!entrada[numEntrada++].equals("FIM"));
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }




  public static void main(String[] args) throws Exception {
    // array que armazena as linhas de entrada
    int[] entrada = new int[maxEntradas];

    // Primeira leitura da entrada padrao com os ids
    int numEntrada = leituraDeEntrada(entrada);

    // formar array de jogadores    
    Lista lista = new Lista();
    
    for(int i = 0; i < numEntrada; i++) {
      Jogador jogador = new Jogador(leLinhaCSV(entrada[i]));
      lista.inserirFim( jogador );
    }

    // segunda leitura de entrada
    int numOperacoes = MyIO.readInt();
    char operacao;

    for (int i = 0; i < numOperacoes; i++) {
      operacao = MyIO.readChar();
      MyIO.readChar();
      switch(operacao) {
        //case "I*": lista.inserir(MyIO.readInt(), new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        //case "II": lista.inserirInicio(new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        case 'I': lista.inserirFim(new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        //case "R*": lista.remover(MyIO.readInt()); break;
        //case "RI": lista.removerInicio(); break;
        case 'R': lista.removerFim(); break;
        default: System.out.println("operacao invalida"); break;
      }
    }

    lista.mostrar();

  }
}
